### DISCUSSION TITLE
issue: TypeScript type for `touchedFields` is invalid when field type is Array

### DISCUSSION THREAD

**Discussion by Aerophite** [react-hook-form/react-hook-form#9591] [Q&A]
### Version Number

7.4.1

### Codesandbox/Expo snack

https://codesandbox.io/s/aged-http-3k10bi?file=/src/App.tsx

### Steps to reproduce

1. Click the example code sandbox
2. Open the console, you should see two logs saying "boolean". This is the _actual_ type of `touchedFields`.
3. Hover over any instance of `touchedFields.firstNames` and you'll see the type is _defined_ as `(true | undefined)[] | undefined` when it should really be `boolean | undefined`
    
<img width="894" alt="image" src="https://user-images.githubusercontent.com/6826088/207674133-0a7bd704-1926-4c6c-b16e-c394198dcaed.png">

### Expected behaviour

The type of a touchedFields key should be `boolean | undefined` regardless of the type of the field being an array.

i.e. If the type is `type Values = { firstNames: Array<string>}` then `touchedFields.firstNames` should be of type `boolean | undefined`

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
```tsx
const {
  field: { onBlur }
} = useController({
  control,
  name: "firstNames"
});
```

a controlled component still represents a single user input instead of an array of inputs. If you want to use FieldArray, you can leverage that by useFieldArray hook

**Reply by Aerophite** [reply]
Although this might be the case for an array of inputs. This is only one field. My specific use case is when you have an `<input type=“file” />`, the value (technically `input.files`) is of type `FileList` which is technically an object containing an array. It is not multiple fields. 

I am using `useController` in order to be able to trigger the `onChange` (passing it an `Array<File>`) and `onBlur` events within a custom component when files are selected for upload.

I _could_ set the value type to `FileList` instead of `Array<File>` but that doesn’t make sense because I then would have to transform the value within every form we want to have files uploaded for. 

Regardless, the issue is that `touchedFields` is _typed_ wrong for TypeScript. The _functionality_ is working as I expect…but because of the invalid typing, I have to cast the type using `touchedFields.files as unknown as boolean | undefined` when I should not have to.

**Reply by Aerophite** [reply]
Here's more proof (I think) that it isn't typed correctly.

https://codesandbox.io/s/condescending-vaughan-xiupjm?file=/src/App.tsx

If, instead of using the `useForm`'s `formState.touchedFields.firstNames`, you look at a `<Controller />` `renderProps.fieldState.isTourched`, you can see that although they both _result_ in the same value, the _type_ for `formState.touchedFields.firstNames` is still `(true | undefined)[] | undefined` (wrong) while the type of `renderProps.fieldState.isTouched` is `boolean` (correct).

So...for me, I'm just going to use the `<Controller />` and use the `renderProps.fieldState.isTouched` instead of `useController` with `formState.touchedFields.firstNames` and that will solve my issue. However, I still believe that there is a typing issue here that needs to be fixed.

**Comment by leapful** [comment]
@Aerophite 

Like @bluebill1049 already explained, `useController` hook and `Controller` component are for single field element, it was never built to be used for array type, you are using a wrong tool for the job.

There is no issue on type definition of `useController` or `Controller`, the prop `fieldState` exists for both `useController` and `Controller` and this props is in type boolean because it was built to check state of single field element.

In you example, you still check `formState.touchedFields` while it can be checked with `fieldState.isTouched` from `useController` too. Note that `fieldState` is the same between `useController` and `Controller`.

**Reply by Aerophite** [reply]
@leapful

We are having a miscommunication. I am _not_ saying that the functionality is wrong. In fact, despite what you say about it only working for single values, it _does_ work for Array fields, you just have to manually call `onChange`/`onBlur` (which I'm fine doing). The functionality of both `useController` and `<Controller />` are working as I expect and returns a single boolean value, as expected, when getting the state for the key. 

What I _am_ saying is that your type definitions for TypeScript _are_ wrong. According to your type definitions, when an Array value is used, `formState.touchedFields` _says_ that the type for the Array value's key is  `(true | undefined)[] | undefined` (incorrect as it is _actually_ `boolean | undefined`) while the `<Controller />` `fieldState.isTouched` _correctly_ has the type as `boolean | undefined`. 

We can argue all day about whether you directly support an Array being used here...but that is not my point at all. I am just trying to report that there _is_ an issue with your type definitions. As far as I'm concerned, I'll just use `<Controller />` instead of `useController` since `<Controller />` has the correct definition.

**Reply by leapful** [reply]
@Aerophite 

The `formState.touchedFields` is for the whole form values and and for case array fields, it has to be an Array<boolean | undefined> as it represents for each field value inside the Array.

The `fieldState.isBoolean` is for single field value so it has to be boolean value. 

https://github.com/react-hook-form/react-hook-form/blob/cf5a81f8452d310d7df37157f25cf07b634a5006/src/useController.ts#L153-L155

The implementation is `!!get(formState.touchedFields, name)` meaning is always boolean no matter what the type is.
